### PR TITLE
pin-spec: make changes based on pin spec discussions of 01 Dec 2021

### PR DESCRIPTION
**Pull request by rsnancollas**

### CODE REVIEW

File: part-spec/definitions.json [hunk 1]
```diff
@@ -1,250 +1,315 @@
...
-                    }
-                }
-            }
+  "$id": "https://github.com/chromeos/digital-datasheets/blob/main/part-spec/definitions.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "definitions": {
+    "pinSpec": {
+      "type": "object",
+      "required": [
+        "number",
```
Comment (rsnancollas) [inline part-spec/definitions.json]: terminalIdentifier

File: part-spec/definitions.json [hunk 2]
```diff
@@ -1,250 +1,315 @@
...
+        },
+        "name": {
+          "description": "name given to the signal appearing at the terminal of a component",
+          "type": "string"
+        },
+        "standardized_name": {
+          "description": "standard name of pin",
+          "type": "string",
+          "enum": [
+            "d",
```
Comment (rsnancollas) [inline part-spec/definitions.json]: drain, gate, source, etc.

File: part-spec/definitions.json [hunk 3]
```diff
@@ -1,250 +1,315 @@
...
+          "description": "name of the function of a pin",
+          "examples": [
+            "UART5_TX",
+            "SPI3_MOSI",
+            "PC12"
+          ],
+          "type": "string"
+        },
+        "interfaceType": {
+          "description": "type of interface enabled by pin",
```
Comment (rsnancollas) [inline part-spec/definitions.json:229]: function

File: part-spec/definitions.json [hunk 4]
```diff
@@ -1,250 +1,315 @@
...
+            "usb4",
+            "pcie",
+            "sdio",
+            "memory",
+            "emmc",
+            "annode",
+            "cathode"
+          ],
+          "type": "string"
+        },
```
Comment (rsnancollas) [inline part-spec/definitions.json:245]: pinUsage: 

"enum" : [
UART_TX
UART_RX
SPI_COPI
etc.


]

File: part-spec/definitions.json [hunk 5]
```diff
@@ -1,250 +1,315 @@
...
+        "direction": {
```
Comment (rsnancollas) [inline part-spec/definitions.json]: split this into
"direction"
"electrical characteristics - come up with a better name than this!"

input, push-pull, open-drain,

File: part-spec/definitions.json [hunk 6]
```diff
@@ -1,250 +1,315 @@
...
+          "description": "absolute minimum voltage that can be applied to a pin",
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
+        "vmax": {
+          "description": "maximum continuous voltage that can safely be applied to a pin",
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
+        "imax": {
```
Comment (rsnancollas) [inline part-spec/definitions.json:93]: imin (current that can be sunk)

File: part-spec/definitions.json [hunk 7]
```diff
@@ -1,250 +1,315 @@
...
+        "configuration"
+      ],
+      "properties": {
+        "type": {
+          "description": "type of component connected to a pin",
+          "type": "string",
+          "enum": [
+            "resistor",
+            "capacitor",
+            "fet",
```
Comment (rsnancollas) [inline part-spec/definitions.json]: maybe get rid of fet/diode

File: part-spec/definitions.json [hunk 8]
```diff
@@ -1,250 +1,315 @@
...
+          "description": "list of voltage levels supported by a pin",
+          "comment": "units of volts",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/unit"
+          }
+        },
+        "floatUnused": {
+          "description": "description of whether pin can safely be floated if it is not used",
+          "type": "boolean"
```
Comment (rsnancollas) [inline part-spec/definitions.json:123]: esd note:
- protection exists
- maybe also indicate which rail it's pulled to

Comment (rsnancollas) [inline part-spec/definitions.json:123]: internal pull

File: part-spec/definitions.json [hunk 9]
```diff
@@ -1,250 +1,315 @@
...
+        "type": {
+          "description": "type of component connected to a pin",
+          "type": "string",
+          "enum": [
+            "resistor",
+            "capacitor",
+            "fet",
+            "diode"
+          ]
+        },
```
Comment (rsnancollas) [inline part-spec/definitions.json:162]: also indicate rail to connect to

File: part-spec/definitions.json (hunk 1 above)
Comment (rsnancollas) [reply part-spec/definitions.json]: done

File: part-spec/definitions.json (hunk 2 above)
Comment (rsnancollas) [reply part-spec/definitions.json]: done

File: part-spec/definitions.json [hunk 10]
```diff
@@ -1,250 +1,339 @@
...
+          "description": "voltage ouput low or minimum voltage of a high output",
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
+        "voh": {
+          "description": "voltage ouput high or maximum voltage of a high output",
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
+        "abs_vmax": {
```
Comment (asadate) [inline part-spec/definitions.json]: absVmax to be consistent

File: part-spec/definitions.json [hunk 11]
```diff
@@ -1,250 +1,339 @@
...
+          "description": "absolute maximum voltage that can be applied to a pin",
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
+        "abs_vmin": {
```
Comment (asadate) [inline part-spec/definitions.json]: "absVmin"

File: part-spec/definitions.json [hunk 12]
```diff
@@ -1,250 +1,339 @@
...
+      "properties": {
+        "terminalIdentifier": {
+          "description": "pin or ball number as defined by datasheet",
+          "type": "string"
+        },
+        "name": {
+          "description": "name given to the signal appearing at the terminal of a component",
+          "type": "string"
+        },
+        "standardized_name": {
```
Comment (asadate) [inline part-spec/definitions.json]: "standardizedName" to be consistent

File: part-spec/definitions.json [hunk 13]
```diff
@@ -1,250 +1,339 @@
...
+        },
+        "internalPullUp": {
+          "description": "indicates the value of an internal pull-up on a pin",
+          "comment": "units of resistance",
+          "$ref": "#/definitions/unit"
+        },
+        "internalPullDown": {
+          "description": "indicates the value of an internal pull-down on a pin",
+          "comment": "units of resistance",
+          "$ref": "#/definitions/unit"
```
Comment (asadate) [inline part-spec/definitions.json:123]: I can no longer find the Unit definition section in this updated version. Was it deleted?

File: part-spec/definitions.json [hunk 14]
```diff
@@ -1,250 +1,339 @@
...
+        },
+        "functionProperities": {
+          "description": "list of function objects that can apply to an individual pin",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/pinFunction"
+          }
+        },
+        "vihMin": {
+          "description": "minimum voltage at which an input is read as high",
```
Comment (asadate) [inline part-spec/definitions.json]: JEDEC definition of vihMin: The least positive (most negative) value of high-level input voltage for which operation of the logic element within specification limits is to be expected.

File: part-spec/definitions.json [hunk 15]
```diff
@@ -1,250 +1,339 @@
...
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
+        "vilMax": {
+          "description": "maximum voltage at which an input is read as low",
```
Comment (asadate) [inline part-spec/definitions.json]: JEDEC definition of vilMax: The most positive (least negative) value of low-level input voltage for which operation of the logic element within
specification limits is to be expected

File: part-spec/definitions.json [hunk 16]
```diff
@@ -1,250 +1,339 @@
...
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
+        "vol": {
+          "description": "voltage ouput low or minimum voltage of a high output",
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
```
Comment (asadate) [inline part-spec/definitions.json:72]: We should probably add vihMax: The most positive (least negative) value of high-level input voltage for which operation of the logic element within specification limits is to be expected.
And vilMin: The least
positive (most negative) value of low-level input voltage for which operation of the logic element within specification limits is to be expected.

File: part-spec/definitions.json [hunk 17]
```diff
@@ -1,250 +1,339 @@
...
+        "vilMax": {
+          "description": "maximum voltage at which an input is read as low",
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
+        "vol": {
+          "description": "voltage ouput low or minimum voltage of a high output",
```
Comment (asadate) [inline part-spec/definitions.json]: JEDEC vol: The voltage level at an output terminal with input conditions applied that, according to the product specification, will establish a low level at the output

File: part-spec/definitions.json [hunk 18]
```diff
@@ -1,250 +1,339 @@
...
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
+        "voh": {
+          "description": "voltage ouput high or maximum voltage of a high output",
```
Comment (asadate) [inline part-spec/definitions.json]: JEDEC voh: The voltage level at an output terminal with input conditions applied that, according to the product specification, will establish a high level at the output.

File: part-spec/definitions.json [hunk 19]
```diff
@@ -1,250 +1,339 @@
...
+          "comment": "units of volts",
+          "$ref": "#/definitions/unit"
+        },
+        "abs_vmax": {
+          "description": "absolute maximum voltage that can be applied to a pin",
+          "comment": "units of volts",
```
Comment (asadate) [inline part-spec/definitions.json:80]: absVmax: Maximum voltage rating beyond which damage to the device may occur

File: part-spec/definitions.json [hunk 20]
```diff
@@ -1,250 +1,339 @@
...
+        "description": {
+          "description": "description of the signal appearing at the terminal of an electric/electronic component",
+          "type": "string"
+        },
+        "numberOfSupportedFunctions": {
+          "description": "the total number of functions supported by this pin",
+          "type": "number"
+        },
+        "functionProperities": {
+          "description": "list of function objects that can apply to an individual pin",
```
Comment (asadate) [inline part-spec/definitions.json:42]: Typo: Should be functionProperties

File: part-spec/definitions.json [hunk 21]
```diff
@@ -1,250 +1,339 @@
...
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/pinFunction"
+          }
```
Comment (asadate) [inline part-spec/definitions.json:46]: pinFunction is missing from the bottom of this updated version

File: part-spec/definitions.json [hunk 22]
```diff
@@ -1,250 +1,339 @@
...
+      }
+    },
+    "externalComponents": {
+      "type": "object",
+      "required": [
+        "type",
+        "configuration"
+      ],
+      "properties": {
+        "type": {
```
Comment (asadate) [inline part-spec/definitions.json]: Looking at this again, "type" here can be confusing with  "type" we use for data type. Change this property to "componentType"?

File: part-spec/definitions.json [hunk 23]
```diff
@@ -1,250 +1,339 @@
...
+        	"example": "multiply",
+        	"type": "string",
+        	"enum": [
+						"multiply",
+						"add",
+						"subtract"
+						],
+      },
+      "tbd": {
+      	"description": "a boolean representing whether a value has been defined",
```
Comment (asadate) [inline part-spec/definitions.json:345]: Do we want to call "tbd" something like 'valueDefined" to match description?

File: part-spec/definitions.json (hunk 18 above)
Comment (rsnancollas) [reply part-spec/definitions.json]: adjusted vol as well

**Review by asadate — APPROVED** [review]

**Pull request merged**